src/api/java/com/minecolonies/api/crafting/RecipeStorage.java

```diff
@@ -382,22 +383,29 @@ public boolean equals(final Object o)
     @Override
     public int hashCode()
     {
-        int result = cleanedInput.hashCode();
-        result = 31 * result + primaryOutput.hashCode();
-        result = 31 * result + (intermediate != null ? intermediate.hashCode() : 0);
-        result = 31 * result + gridSize;
-        if(recipeSource != null)
-        {
-            result = 31 * result + recipeSource.hashCode();
-        }
-        if(recipeType != null && !(recipeType instanceof ClassicRecipe))
+        return Objects.hash(cleanedInput, 
+            primaryOutput.getItem(),
+            primaryOutput.getCount(),
+            intermediate,
+            gridSize,
+            hashableItemStackList(alternateOutputs),
+            hashableItemStackList(secondaryOutputs),
+            hashableItemStackList(tools));
+    }
+
+    /**
+     * Convert a list of itemstacks into something hashable
+     * @param items List of item stacks to convert
+     * @return hashtable of items and counts
+     */
+    private Map<Item, Integer> hashableItemStackList(List<ItemStack> items)
+    {
+        Map<Item, Integer> hashableList = new HashMap<>();
```

Comment:
seeing as the hashing now seems to get pretty extensive, does the Recipestorage ever change? if not  we could just calc it once

Reply:
theoretically that should be possible. But afaik that's "bad style". Unless the object really is immutable

Reply:
well we're using it relatively frequently right

Reply:
er actually it gotta keep the same hash, else we run into issues with the map thinking we got the old hash while we dont

Reply:
The one inside the map yes, the ones we push around outside not

Reply:
RecipeStorages are effectively immutable. All of their settings happen in the constructor. Caching should be safe.